### ISSUE TITLE
Need help with Java + API

### ISSUE THREAD

**Issue by kkingsley-BF** [PrivateBin/PrivateBin#423]
I've been racking my brain over this issue for the last few days and am not making any progress. I am trying to utilize the API to receive a URL that I would normally get by using the web UI. This is the Java code I have pieced together. It will generate the salt, iv, ct, key, and password then pass it off to https://privatebin.net via cURL command. I receive back an ID and then create a URL in the form of https://privatebin.net/?{id-received}#{secret-key}. When I open the URL in the browser it asks for a password. I have tried entering the password the script generates with no luck. I have also tried various combinations with/without base64 padding on the salt/iv/ct/key/password.

I'm thinking there is an error in the way I am generating the data that is passed to the API but I am at a loss. Any help would be appreciated. 

```java
import java.util.concurrent.ThreadLocalRandom
import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.spec.KeySpec;
import groovy.json.JsonSlurper

String plainTextMessage = "Hello from Jira"

//Generate password
KeyGenerator keyGen = KeyGenerator.getInstance("AES");
keyGen.init(32);
String password = Base64.getEncoder().withoutPadding().encodeToString(keyGen.generateKey().getEncoded());

//Generate Salt
Random random = ThreadLocalRandom.current();
byte[] salt = new byte[8];
random.nextBytes(salt);
String saltString = Base64.getEncoder().withoutPadding().encodeToString(salt);

//Generate Secret Key
SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
KeySpec passwordBasedEncryptionKeySpec = new PBEKeySpec(password.toCharArray(), salt, 1000, 256);
SecretKey tmp = factory.generateSecret(passwordBasedEncryptionKeySpec);
SecretKey secret = new SecretKeySpec(tmp.getEncoded(), "AES");
String secretString = Base64.getEncoder().encodeToString(secret.getEncoded())

//Generate IV
byte[] iv = new byte[16];
random.nextBytes(iv);
String ivString = Base64.getEncoder().withoutPadding().encodeToString(iv);

//Generate Cipher
Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
GCMParameterSpec spec = new GCMParameterSpec(128, iv);
cipher.init(Cipher.ENCRYPT_MODE, secret, spec);
byte[] cipherTextBytes = cipher.doFinal(plainTextMessage.getBytes(StandardCharsets.UTF_8));
String cipherText = Base64.getEncoder().withoutPadding().encodeToString(cipherTextBytes);

//Curl Command
String header = 'X-Requested-With: JSONHttpRequest'
String hostURL = 'https://privatebin.net'
String curlCommand = "curl -k -X POST -H \"${header}\" \"${hostURL}\" --data 'data={\"iv\":\"${ivString}\",\"v\":1,\"iter\":1000,\"ks\":256,\"ts\":128,\"mode\":\"gcm\",\"adata\":\"\",\"cipher\":\"aes\",\"salt\":\"${saltString}\",\"ct\":\"${cipherText}\"}&expire=never&formatter=plaintext&burnafterreading=1&opendiscussion=1'"
Process process = ["/bin/sh", "-c", "${curlCommand}"].execute()
def out = new StringBuffer()
def err = new StringBuffer()
process.consumeProcessOutput( out, err )
process.waitFor()

//Final output
String output = out
def outputJSON = new JsonSlurper().parseText(output)
if (outputJSON["status"] == 0){
    String privateBinURL = "URL: ${hostURL}${outputJSON["url"]}#${secretString}\nPassword: ${password}"
    return privateBinURL
}
if (outputJSON["status"] == 1){
    return outputJSON["message"]
}
```

**Comment by rugk** [comment]
You know you also need to compress the data? Especially, also notice the current web implementation uses a faulty zlib implementation, see https://github.com/PrivateBin/PrivateBin/pull/193.

**Comment by elrido** [comment]
Please have a look at the recently updated [encryption format wiki page](https://github.com/PrivateBin/PrivateBin/wiki/Encryption-format). If you still have questions, feel free to reopen the issue.

**Comment by kkingsley-BF** [comment]
Thanks for sending the updated info. I was able to finalize this script this weekend. I had basically given up on it once I found out about the faulty zlib stuff. Since v1.3 corrects this issue I picked up where I left off. So for anyone who stumbles across this threat here's a groovy script that is fully functional and brings every feature of PrivateBin to groovy. If anything, it may help others to figure out how to write a script on their own in a different language.

Our use case for this script will be to utilize it with Jira using ScriptRunner add on to deliver secrets.

I've created a public repo for the script. Do as you wish with it.
https://github.com/kkingsley-BF/PrivateBin-Groovy/blob/master/Paste.groovy
